src/neo/SmartContract/ApplicationEngine.cs

```diff
@@ -85,18 +85,49 @@ private StackItem ConvertReturnValue(object value)
                 uint i => i,
                 long i => i,
                 ulong i => i,
-                Enum e => ConvertReturnValue(Convert.ChangeType(e, e.GetTypeCode())),
+                Enum e => Convert(System.Convert.ChangeType(e, e.GetTypeCode())),
                 byte[] data => data,
                 string s => s,
-                UInt160 i => i.ToArray(),
-                UInt256 i => i.ToArray(),
+                BigInteger i => i,
                 IInteroperable interoperable => interoperable.ToStackItem(ReferenceCounter),
-                IInteroperable[] array => new VMArray(ReferenceCounter, array.Select(p => p.ToStackItem(ReferenceCounter))),
+                ISerializable i => i.ToArray(),
                 StackItem item => item,
+                (object a, object b) => new Struct(ReferenceCounter) { Convert(a), Convert(b) },
+                Array array => new VMArray(ReferenceCounter, array.OfType<object>().Select(p => Convert(p))),
                 _ => StackItem.FromInterface(value)
             };
         }
 
+        internal object Convert(StackItem item, InteropParameterDescriptor descriptor)
+        {
+            if (descriptor.IsArray)
+            {
+                Array av;
+                if (item is VMArray array)
+                {
+                    av = Array.CreateInstance(descriptor.Type.GetElementType(), array.Count);
+                    for (int i = 0; i < av.Length; i++)
+                        av.SetValue(descriptor.Converter(array[i]), i);
+                }
+                else
+                {
+                    av = Array.CreateInstance(descriptor.Type.GetElementType(), (int)item.GetBigInteger());
```

Comment:
Does we need this? it could produce an `OutOfMemoryException`, I think that we should provide always an array because it's limited by the VM

Reply:
It is used by `CheckMultisig`.

Reply:
Maybe we should remove this option and allow only one way

Reply:
I've fixed it in the last commit. Please check it.